### DISCUSSION TITLE
How to use sparseblockStructure?

### DISCUSSION THREAD

**Discussion by ErenC1** [UvaCsl/HemoCell#106] [General discussion]
Hi,

I am having trouble with parallelization due to my "irregular" shaped domains and looks like sparseblock is my solution. (HemoCell 2.6)
https://github.com/UvaCsl/HemoCell/blob/d91698721f7a06166cb6ac8d22fa2f2f4baa4ed7/core/hemoCell.cpp#L538-L542
https://github.com/UvaCsl/HemoCell/blob/d91698721f7a06166cb6ac8d22fa2f2f4baa4ed7/core/hemoCell.cpp#L559
What I understood from these lines code checks the config.xml file for mABx,y,z and if it is found, code uses these values to overwrite regulardistribution. I don't need to do any change in my *case*.cpp right? like adding some lines to             voxelizedDomain.get()->getMultiBlockManagement(**HERE**), it is already in the hemocell.cpp file.

I inserted mABx,y,z lines (separately) to config.xml/domain
But when I run the case I don't see the log line that indicates domain decomposition is overwritten, neither domain is decomposed by using mABx,y,z. What could be wrong here?

**Comment by gzavo** [comment]
Hi @ErenC1,

For the moment HemoCell does not support sparse structures. The mAB{x,y,z} parameters help overwrite the atomic block size.
By default we use the recursive (and brute force) domain subdivision of Palabos, but at some specific domain sizes it can create strange atomic block (e.g., very flat in one dimension). In such a case these parameters allow overwriting the default atomic block size. Note however, that in this case, you need to make sure you will end up with the number of atomic blocks you want (e.g. =to cpu core count).

In the future the aim is to also use this parameter to define sparse structures. In theory this should not be a very difficult problem, since Palabos and all the data structures used in HemoCell should support functionality for this out of the box. (Then again, in theory...).

If you are interested, look into the Palabos Tutorial 2.4, and the use of the function "reparallelize".

**Comment by ErenC1** [comment]
Hi @gzavo 

I tried to overwrite standard decomposition when preInlet.h is present, in that case, mAB(x,y,z) lines read from the config file and I could change the structure.
I actually looked at that example (reparallelize) before starting this topic but I couldn't understand some of the lines, like What is that envelope? it is like tolerance for decomposition? etc. so I asked here. But since it is working with preInlet, I have no problem. Thank you for your answer.
                createRegularDistribution3D(
                    parameters.getNx(), parameters.getNy(), parameters.getNz(),
                    global::mpi().getSize(), 1, 1),  // envelopeWidth
                defaultMultiBlockPolicy3D().getThreadAttribution(), 1),
